cpp/src/io/parquet/page_enc.cu

```diff
@@ -200,21 +205,21 @@ __global__ void __launch_bounds__(128)
                        device_2dspan<PageFragment const> fragments,
                        device_span<parquet_column_device_view const> col_desc)
 {
-  // TODO: why not 1 block per warp?
-  __shared__ __align__(8) statistics_group group_g[4];
-
-  uint32_t lane_id              = threadIdx.x & 0x1f;
-  uint32_t frag_id              = blockIdx.y * 4 + (threadIdx.x >> 5);
-  uint32_t column_id            = blockIdx.x;
-  auto num_fragments_per_column = fragments.size().second;
-  statistics_group* const g     = &group_g[threadIdx.x >> 5];
-  if (!lane_id && frag_id < num_fragments_per_column) {
-    g->col       = &col_desc[column_id];
-    g->start_row = fragments[column_id][frag_id].start_value_idx;
-    g->num_rows  = fragments[column_id][frag_id].num_leaf_values;
+  uint32_t const lane_id                  = threadIdx.x & 0x1f;
```

Comment:
There are some rather innocuous-seeming magic values that all mean `32 == 0x1f == cudf::detail::warp_size` in this function. I'll point them out, but I am fine with doing nothing if we feel the current code is better not to change.
```suggestion
  uint32_t const lane_id                  = threadIdx.x % cudf::detail::warp_size;
```

Reply:
sounds good (although using the mod operator makes my teeth itch :rofl:).  Does anyone happen to know if there are constants anywhere for the max threadblock dimensions?  Or are those per-card values?

Reply:
I believe there are no constants for that, and that's why we defined `cudf::detail::warp_size`. It is a constant for all NVIDIA GPUs as far as I am aware.

These two snippets should compile out roughly the same. Compilers can recognize that unsigned modulo by $2^N$ is equivalent to bitwise-and with $2^N - 1`$. Evidence: https://godbolt.org/z/r4c41va5P

Reply:
I went ahead and added a constexpr for the warp mask...there are several other instances of 0x1f sprinkled about in this file that can be replaced later.

Thanks for the link @bdice!

Reply:
To be a bit more precise here, CUDA does provide `warpSize`, which is available inside device code, and the `getDeviceProperties` host function, which returns a struct containing the warp size. However, neither of them is a constant and therefore cannot be used in constant expressions (e.g. for declaring a C-style or std::array). [The warp size is indeed constant across all current compute capabilities](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#features-and-technical-specifications__technical-specifications-per-compute-capability). In theory that's not something that we promise, so the _technically_ correct answer is that we can't use a compile-time constant because in theory someone could run on a new architecture with a different answer. In practice, NVIDIA has no plans to change the warp size AFAIK and many examples of GPU code (even lots of code written by NVIDIA) define a `warp_size` constant. Lots of places use it assuming that it is in fact a compile-time constant and would have to be rewritten if we ever had any cards with a different warp size, so that's a much bigger problem to deal with another day if that ever changes :)